### PR TITLE
Fix Issue 6317

### LINKED ISSUE #6317: Spurious resource leak report when returning before a variable's definition
Given this code:
```java
import java.io.*;
import org.checkerframework.checker.nullness.qual.*;
import org.checkerframework.checker.mustcall.qual.*;
import org.checkerframework.checker.calledmethods.qual.*;

abstract class TestCase {

  abstract @Nullable Closeable alloc();
  abstract boolean arbitraryChoice();

  void method() throws IOException {

    if (arbitraryChoice()) {
      return;
    }

    Closeable r1 = alloc();
    if (r1 == null) {
      return;
    }
    r1.close();

  }

}
```

The RLC reports
```
error: [required.method.not.called] @MustCall method close may not have been invoked on r1 or any of its aliases.
      Closeable r1 = alloc();
                ^
  The type of object is: java.io.Closeable.
  Reason for going out of scope: regular method exit
```

The error looks spurious. Indeed, removing the early return block
```java
      if (arbitraryChoice()) {
        return;
      }
```
silences the warning! It appears that the RLC wants us to close `r1` at the premature return site, before it has been defined.

### PR DESCRIPTION
**Pull request by msridhar**
Fixes #6317.  We try to use the specific store for a control-flow edge after a conditional block to get more precise results.  Also fix a bug in the handling of `@CalledMethodsBottom` when checking must-call obligations.

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java [hunk 1]
```diff
@@ -2121,10 +2122,13 @@ private void propagateObligationsToSuccessorBlock(
...
-              obligationGoesOutOfScopeBeforeSuccessor
-                  ? analysis.getInput(currentBlock).getRegularStore() // 1a. (CM)
-                  : regularStoreOfSuccessor; // 1b. (CM)
+          cmStore = getStoreForEdgeFromEmptyBlock(currentBlock, successor); // 1b. (CM)
+          // TODO we would like to get the store for the specific control-flow edge that the
+          // Must Call Checker computed as well.  However, this does not seem to be possible with
+          // current APIs.  The MustCallAnnotatedTypeFactory keeps a flowResult field that
+          // internally holds the relevant TransferInput objects in its stores map, but it does not
+          // expose these objects via its API.  TransferInput objects keep the "then store" and
+          // "else store" information that we need
```
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:2131]: Specifically, the APIs we need in `AnalysisResult` would be something like `getThenStoreBefore(Block)` and `getElseStoreBefore(Block)`, so we could write code similar to what is written starting at line 2184.  @smillst @mernst do you think these would be reasonable APIs to add?

Comment (msridhar) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java]: I decided to just punt on this for now, as I cannot think of a realistic test case that expose any imprecision.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java [hunk 2]
```diff
@@ -2095,16 +2097,19 @@ private void propagateObligationsToSuccessorBlock(
...
+        // store propagated along the CFG edge from currentBlock to successor.  But, there are
+        // special cases to consider.  The rules are:
+        // 1. if the current block has no nodes, it is either a ConditionalBlock or a SpecialBlock.
+        //    For the called-methods store, we obtain the exact CFG edge store that we need (see
+        //    getStoreForEdgeFromEmptyBlock()).  For the must-call store, due to API limitations,
+        //    we use the following heuristics:
         //    1a. if there is information about any alias in the resource alias set
-        //        in the successor store, use the successor's CM and MC stores, which
+        //        in the successor store, use the successor's MC store, which
         //        contain whatever information is true after this block finishes.
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:2110]: typo: "contains"

**Review by kelloggm — DISMISSED** [review]
Nice fix, LGTM. I don't need to review again after the typo is fixed.

**Comment by msridhar** [comment]
@kelloggm could you stamp again?  The typo fix dismissed your review.

Also, I turned off auto-merge as it looks like a release is currently happening.

**Review by kelloggm — APPROVED** [review]

**Pull request merged**